### PR TITLE
Auto-add needed tech when adding an item

### PR DESCRIPTION
**Pull request by pikiou**
When adding e.g. Broodlords in an empty bo, it will add all the required structures and tech needed for it. Works for upgrades and structures as well.
Will do the same for terran and protoss next if this pull request is merged.

### CODE REVIEW

**Comment by pikiou** [comment]
New files weren't added to git because I used `git commit -am "explanation"` when it should have been `git add -A && git commit -m "explanation"`
Added auto-add for protoss. TechLab and Reactor give me a harder time for terran.
Added a new menu button like Settings, but for optimizations on the build. There is a lot to add here it'll be a ton of fun!

**Comment by jtsauls** [comment]
Awesome looking updates!

**Comment by BurnySc2** [comment]
Thank you.

But let me slow you down, I want to merge this PR first before it gets too complicated and I am unable to follow the changes.
I hope that is alright with you. If not, please add a `[WIP]` (for work in progress) in front of the PR title - maintainers should usually know that they shouldn't merge those PRs.

**Comment by BurnySc2** [comment]
1)
I found a "bug" where if you remove an item after optimizing the build order by workers, the "selection" drops one to the left
So from 
image
it goes to 
image
while the position of the "selection" shouldn't change.
However I am currently unable to reproduce the bug.

2) 
A suggestion would be to hide the "Optimize" drop-down for races that are not working at the moment (so terran right now).
If you can't add it, I could add it later - I won't have much time until monday evening.

**Comment by pikiou** [comment]
1\)
I think it was happening before from my commits.
Here is an example on https://burnysc2.github.io/sc2-planner/?&race=protoss&bo=002eJyLrlbKTFGyMjPQUSqpLEhVslIqLikqTS4pLUpVqtWBShrhkTS1xCNpYQKXLM8vyk4tIkLG3BguU5qXWUKRWSTJxAIA32hVIA==
If you place the selection say after the nexus and you double click on the zealot, hence removing the zealot and the following probe, it will move the selection one square to the left.
Screenshot from 2021-05-07 02-14-42
Or are you referring to another bug?

2\)
I didn't know my commits were added to this pull request, so there are multiple features in this one but from now on I'll create branches to prevent this. Sorry!
The features are :
- when adding an item to the build order, it also adds all its requirements (it doesn't work for terran but I'm working on it)
- when clicking on the Optimize button, a drop down menu offers various optimisations. The first one is to add as many workers as possible as long as the build order doesn't end later
- by default all the build order is visible, and a button allows to limit it to one scrollable line.

I'll try to maintain [this gh page](https://pikiou.github.io/sc2-planner) up to date so you can test out these features before merging them.

**Comment by pikiou** [comment]
This is ready to be pulled (do I need to comment about it or if changing its title was sufficient?)

**Comment by BurnySc2** [comment]
> If you place the selection say after the nexus and you double click on the zealot, hence removing the zealot and the following probe, it will move the selection one square to the left.

Yes, exactly - very weird 'bug'. I'll try to check out what causes it or how to fix it.

> I'll try to maintain this gh page up to date so you can test out these features before merging them.

I'll pull the PR and `npm start` it on my end to try it out anyway - thanks though!

File: src/components/Optimize.tsx
Comment (BurnySc2) [inline src/components/Optimize.tsx:58]: Here you are just converting number to string, I think
Perhaps format string would make it easier to read:
```this.props.applyOpitimization([`${optimizationName}`])```

**Review by BurnySc2 — APPROVED** [review]
Thank for adding the tests.
Tests are always very valueable and welcome.

I tried to understand the changes but I don't think I was able to follow all of them. Otherwise your coding style looks good.

Edit:
Could you point out where the code might slow down?
Compare the following two:
<https://burnysc2.github.io/sc2-planner/?&race=terran&bo=002eJzFVEEKgzAQ/EvOHpo1YupXSg+iUkJBJU0opfTvvZmLO5pU0usOs5mZ3ezlLUwvGlUVwr3mQTTiOdn7YMWn2ELkeUEezvrOeTvsoPEISdCQTkmv0QK0nTPT+JNAWRISEfLwo3GBpQEJhljHi4cZlkhJYsBJiF5Piq2zncoaOeLiI8TaGPJh0vnhRrfKIEoGip9vtu13RHL05SC0vgqu72EBZTP7T8kJdwe6xKBmd05Ff5Foq5VaZ3D1LGHG2iN0sSCYbdHXwesXmJar9Q==>

<https://pikiou.github.io/sc2-planner/?&race=terran&bo=002eJzFVEEKgzAQ/EvOHpo1YupXSg+iUkJBJU0opfTvvZmLO5pU0usOs5mZ3ezlLUwvGlUVwr3mQTTiOdn7YMWn2ELkeUEezvrOeTvsoPEISdCQTkmv0QK0nTPT+JNAWRISEfLwo3GBpQEJhljHi4cZlkhJYsBJiF5Piq2zncoaOeLiI8TaGPJh0vnhRrfKIEoGip9vtu13RHL05SC0vgqu72EBZTP7T8kJdwe6xKBmd05Ff5Foq5VaZ3D1LGHG2iN0sSCYbdHXwesXmJar9Q==>
Try to notice the delay/lag, if you add an SCV at the end. In your version, it seems to take about double as long. Do you have an idea where the performance hit could come from?
Btw, there are huge performance differences between `npm run start` and `npm run build` versions.

File: src/components/Optimize.tsx
Comment (pikiou) [reply src/components/Optimize.tsx]: I changed it, thanks for the advice!

**Comment by pikiou** [comment]
The performance issue mostly came from JSON.stringify being called about 100000 times per bo simulation as an error message detail in the idleUnits loop from trainUnit() in GameLogic
That was hell to spot.

Indeed the auto-add requirements feature only works if you add something at the end of the build. It could work everywhere now that I understand the project, I'll have a look at the possibilities but it's less exciting that new features I'm working on...

All the builds I created were on one base, I didn't notice problems regarding timings. My next builds will be on 2 bases I might need accurate income estimations! So I might fix it before you do :P

**Comment by pikiou** [comment]
I have a few pull requests waiting for this one to be merged. You may have a lot on your plate already, or maybe it's not good enough ><
Again I'm new to github maybe should I just wait. Please don't hesitate to tell me how things are done around here !

**Comment by BurnySc2** [comment]
Sorry, it's good.
I just need a couple more days to merge it

**Pull request merged**
